R/estimateDiversity.R

```diff
@@ -149,23 +158,30 @@ NULL
 
 #' @rdname estimateDiversity
 #' @export
-setGeneric("estimateDiversity",signature = c("x"),
+setGeneric("estimateDiversity",signature = c("x", "tree"),
            function(x, abund_values = "counts",
                     index = c("shannon","gini_simpson", "inverse_simpson",
-                              "coverage", "fisher"),
-                    name = index, ...)
+                              "coverage", "fisher", "pd"),
+                    name = index, tree = rowTree(x), ...)
                standardGeneric("estimateDiversity"))
 
 #' @rdname estimateDiversity
 #' @export
-setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
+setMethod("estimateDiversity", signature = c(x="SummarizedExperiment", tree="missing"),
     function(x, abund_values = "counts",
              index = c("shannon","gini_simpson","inverse_simpson",
                        "coverage", "fisher"),
-             name = index, ..., BPPARAM = SerialParam()){
+             name = index, tree = "missing", ..., BPPARAM = SerialParam()){
 
         # input check
+        # IF object does not have a tree
+        if( "pd" %in% index ){
+            stop("SummarizedExperiment object does not have a tree. 'PD' is not possible to calculate.",
+                 call. = FALSE)
+        }
+
```

Comment:
On the one hand you set the default and on the other hand you don't allow it. The user cannot run the function with the default setting, even though a SE is provided.

This can only be the one or the other and you have to choose. See comment above

Reply:
@FelixErnst do you mean that we should have different exported functions for alpha diversities that use the tree (TSE) and those that do not require the tree (SE)? Or is this discussion about the internal functions and implementation but user could still call all these functions through `estimateDiversity`?

Reply:
Maybe that the right direction to head in. However, I had something different in mind:

- `estimateDiversity` for `SE` without any mention of `pd`
- `estimatePD` for `SE` and `tree`
- `estimateDiversity` for `TSE` dynamically including `pd` if `rowTree` is not `NULL`

With this the contracts from the objects slots are resolved stringently without any including any ambiguous combination of defaults and ability to calculate certain indices.

For even more clarity and easier implementation `pd` can also be omitted for `TSE` by default. This could then be easily implemented like this 


```r
setGeneric("estimatePD", ...)
setMethod("estimatePD", signature = c("SummarizedExperiment","phylo"),
                   ...)
setMethod("estimatePD", signature = c("TreeSummarizedExperiment","missing"),
    function(x, ...){
        if(is.null(rowTree(x))){
            stop("...")
        }
        estimatePD(x, rowTree(x), ...)
    }
)

```

Reply:
I am not sure about my estimatePD either. Currently, first all but "pd" are added to colData, and after that "pd" is added. For example, if those indices are already calculated, then 2 warnings come because first warning says "shannon, simpson... already in colData", and the second warning says "pd is already in colData". 

Those 2 warnings might be weird to user, so should that be done differently? 

And should estimatePD be just normal internal function?

Reply:
The refactor the code accordingly. These warnings are a new addition, so I guess there needs to be additional maturation